platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/consensus/AncestorIterator.java

```diff
@@ -0,0 +1,210 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.consensus;
+
+import com.swirlds.platform.internal.EventImpl;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Instant;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.function.Predicate;
+
+/**
+ * This is an iterator over all valid ancestors of a given root event, that are reachable through
+ * valid ancestors. The "valid" ancestors are defined by a lambda predicate passed in to the
+ * constructor.
+ *
+ * <p>It can be used to read or modify the events. This does not implement Collection, because its
+ * only purpose to allow iteration over those events. The iteration is depth first, and backtracks
+ * each time it reaches an invalid event (one for which the predicate returns false).
+ *
+ * <p>It is not threadsafe, and will silently fail without throwing any exceptions if you attempt to
+ * create and use two at the same time, even if they are in the same thread, and even if they are
+ * only used to read and not write. So always create only one at a time, and ensure that you are
+ * done with it before creating another one.
+ *
+ * <p>This returns all ancestors of the root event that are valid. It iterates in an order that
+ * always returns a parent before its child. The root itself is considered to be one of the
+ * ancestors, and is last in the iteration.
+ *
+ * <p>Recursion happens on self parents before other parents. So if there are multiple paths from
+ * the root to an event, it will use the path that stays on line of self parents for as far down as
+ * possible before leaving that line.
+ */
+public class AncestorIterator implements Iterator<EventImpl> {
+    private static final int INITIAL_STACK_SIZE = 300;
+    private final EventVisitedMark mark;
+    /** stack of EventImpl on the path to curr */
+    private final Deque<EventImpl> stackRef = new ArrayDeque<>(INITIAL_STACK_SIZE);
+    /** stack of state */
+    private final Deque<IteratorState> stackState = new ArrayDeque<>(INITIAL_STACK_SIZE);
+    /** stack of selfAncestor */
+    private final Deque<Boolean> stackSelfAncestor = new ArrayDeque<>(INITIAL_STACK_SIZE);
+    /** stack of timeReachedRoot */
+    private final Deque<Instant> stackTime = new ArrayDeque<>(INITIAL_STACK_SIZE);
+    /** the current event reached in the search */
+    private EventImpl curr;
+    /**
+     * a lambda which filters which ancestors are of interest: only each event e for which
+     * valid(e)==true
+     */
+    private Predicate<EventImpl> valid;
+    /**
+     * the time when the event last returned by the iterator's next() first reached the creator of
```

Comment:
I was confused about this last time, and i'm confused again. What does it mean for an event to "reach the creator of root"?

Reply:
not sure, this is original code that has been repackaged 😄